// @looker/components/dist/es2015/Sidebar/SidebarGroup.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import React, { useState } from 'react';
import styled from 'styled-components';
import { Heading } from '../Text';
import { Box } from '../Layout';
import { Icon } from '../Icon';
const InternalSidebarGroup = (_a) => {
    var { children, className, label, showChildren = false } = _a, props = __rest(_a, ["children", "className", "label", "showChildren"]);
    const [isOpen, setOpen] = useState(showChildren);
    const toggle = () => {
        setOpen(!isOpen);
        return false;
    };
    return (React.createElement("section", Object.assign({ className: className }, props),
        React.createElement(SidebarGroupHeading, { onClick: toggle },
            React.createElement("button", { "aria-expanded": isOpen ? 'true' : 'false' },
                label,
                React.createElement(Icon, { size: 20, name: isOpen ? 'CaretUp' : 'CaretDown' }))),
        isOpen && React.createElement(Box, { px: "medium" }, children)));
};
const SidebarGroupHeading = styled(Heading).attrs({
    as: 'h3',
    fontSize: 'small',
    fontWeight: 'semiBold',
}).withConfig({ displayName: "SidebarGroupHeading", componentId: "sc-16gzick" }) `
  button {
    align-items: center;
    all: inherit;
    display: flex;
    justify-content: space-between;
    width: 100%;

    ${Icon} {
      align-self: center;
    }
  }
`;
export const SidebarGroup = styled(InternalSidebarGroup).withConfig({ displayName: "SidebarGroup", componentId: "sc-1g52mwn" }) ``;
