// @looker/components/dist/esm/Form/Fields/Field.js
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __spreadArrays = (this && this.__spreadArrays) || function () {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
};
import React from 'react';
import styled, { css } from 'styled-components';
import omit from 'lodash/omit';
import pick from 'lodash/pick';
import { FlexItem } from '../../Layout/FlexItem';
import { FormControl } from '../FormControl/FormControl';
import { Label } from '../Label/Label';
import { ValidationMessage, } from '../ValidationMessage/ValidationMessage';
export var CustomizableFieldAttributes = {
    labelMargin: 'xsmall',
    labelWidth: '20%',
};
export var fieldPropKeys = [
    'alignLabel',
    'alignValidationMessage',
    'id',
    'label',
    'labelFontSize',
    'labelFontWeight',
    'labelWidth',
    'validationMessage',
];
export var pickFieldProps = function (props) {
    return pick(props, __spreadArrays(fieldPropKeys, ['required', 'className']));
};
export var omitFieldProps = function (props) { return omit(props, fieldPropKeys); };
var RequiredStar = styled(function (props) { return (React.createElement("span", __assign({}, props, { "aria-hidden": "true" }),
    ' ',
    "*")); }).withConfig({ displayName: "RequiredStar", componentId: "sc-ohzx3f" })(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  color: ", ";\n"], ["\n  color: ", ";\n"])), function (props) { return props.theme.colors.semanticColors.danger.darker; });
var handleHorizontalAlignment = function (props) {
    var alignLabel = props.alignLabel, labelWidth = props.labelWidth;
    var width = labelWidth || CustomizableFieldAttributes.labelWidth;
    switch (alignLabel) {
        case 'left':
            return css(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n        text-align: right;\n        width: ", ";\n      "], ["\n        text-align: right;\n        width: ", ";\n      "])), width);
        case 'right':
            return css(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n        text-align: left;\n        width: ", ";\n        margin-right: 0;\n        margin-left: ", ";\n      "], ["\n        text-align: left;\n        width: ", ";\n        margin-right: 0;\n        margin-left: ",
                ";\n      "])), width, function (props) {
                return props.theme.space[CustomizableFieldAttributes.labelMargin];
            });
        case 'bottom':
        case 'top':
        default:
            return '';
    }
};
var getValidationMessageAlignment = function (alignValidationMessage) {
    switch (alignValidationMessage) {
        case 'left':
            return 'right';
        case 'right':
            return 'left';
        case 'bottom':
            return 'top';
        case 'top':
            return 'bottom';
        default:
            return undefined;
    }
};
var FieldComponent = function (_a) {
    var alignValidationMessage = _a.alignValidationMessage, children = _a.children, id = _a.id, label = _a.label, labelFontSize = _a.labelFontSize, labelFontWeight = _a.labelFontWeight, required = _a.required, validationMessage = _a.validationMessage, props = __rest(_a, ["alignValidationMessage", "children", "id", "label", "labelFontSize", "labelFontWeight", "required", "validationMessage"]);
    return (React.createElement(FormControl, __assign({ mb: "xsmall" }, props),
        React.createElement(Label, { htmlFor: id, fontWeight: labelFontWeight, fontSize: labelFontSize },
            label,
            required && React.createElement(RequiredStar, null)),
        React.createElement(FormControl, { alignLabel: getValidationMessageAlignment(alignValidationMessage), mb: "xsmall" },
            React.createElement(FlexItem, null, children),
            validationMessage ? (React.createElement(ValidationMessage, __assign({ ml: alignValidationMessage === 'right' ? 'xsmall' : undefined }, validationMessage))) : null)));
};
export var Field = styled(FieldComponent).withConfig({ displayName: "Field", componentId: "sc-9vj8xo" })(templateObject_4 || (templateObject_4 = __makeTemplateObject(["\n  ", " {\n    ", "\n  }\n"], ["\n  ", " {\n    ", "\n  }\n"])), Label, handleHorizontalAlignment);
var templateObject_1, templateObject_2, templateObject_3, templateObject_4;
