// @looker/components/dist/cjs/Menu/MenuItem/menuItemStyle.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var design_tokens_1 = require("@looker/design-tokens");
var initial = {
    bg: design_tokens_1.palette.white,
    color: design_tokens_1.palette.charcoal600,
    fontSize: 'small',
    fontWeight: 'normal',
    iconColor: design_tokens_1.palette.charcoal300,
    iconSize: 20,
};
var hover = __assign(__assign({}, initial), { bg: design_tokens_1.palette.charcoal100, color: design_tokens_1.palette.charcoal900 });
var current = __assign(__assign({}, hover), { fontWeight: 'bold', iconColor: design_tokens_1.palette.charcoal900 });
exports.defaultMenuItemStyle = {
    current: current,
    hover: hover,
    initial: initial,
    marker: {
        color: design_tokens_1.palette.charcoal900,
        size: 4,
    },
};
