// @looker/components/dist/cjs/Card/Card.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var design_tokens_1 = require("@looker/design-tokens");
var styled_components_1 = __importStar(require("styled-components"));
var layout_1 = require("../Layout/layout");
var cardTransition = function () { return styled_components_1.css(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  ", "\n"], ["\n  ",
    "\n"])), function (props) {
    return props.theme.transitions.durationQuick + " " + props.theme.easings.ease;
}); };
var raised = function (props) {
    return props.raised && styled_components_1.css(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n    box-shadow: ", ";\n\n    &:hover {\n      box-shadow: ", ";\n    }\n  "], ["\n    box-shadow: ", ";\n\n    &:hover {\n      box-shadow: ", ";\n    }\n  "])), function (props) { return props.theme.shadows[1]; }, function (props) { return props.theme.shadows[2]; });
};
exports.Card = styled_components_1.default.div.withConfig({ displayName: "Card", componentId: "sc-uvy610" })(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n  ", "\n  ", "\n\n  transition: border ", ", box-shadow ", ";\n\n  &:hover {\n    border-color: ", ";\n  }\n\n  ", "\n"], ["\n  ", "\n  ", "\n\n  transition: border ", ", box-shadow ", ";\n\n  &:hover {\n    border-color: ", ";\n  }\n\n  ", "\n"])), layout_1.layoutCSS, design_tokens_1.flexbox, cardTransition, cardTransition, function (props) { return props.theme.colors.palette.charcoal400; }, raised);
exports.Card.defaultProps = {
    bg: 'palette.white',
    border: '1px solid',
    borderColor: 'palette.charcoal300',
    borderRadius: 'medium',
    display: 'flex',
    flexDirection: 'column',
    height: '100%',
    minWidth: '200px',
    overflow: 'hidden',
};
var templateObject_1, templateObject_2, templateObject_3;
