// @looker/components/dist/cjs/Button/ButtonGroup.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var uniqueId_1 = __importDefault(require("lodash/uniqueId"));
var xor_1 = __importDefault(require("lodash/xor"));
var react_1 = __importStar(require("react"));
var styled_components_1 = __importDefault(require("styled-components"));
var utils_1 = require("../utils");
var ButtonItem_1 = require("./ButtonItem");
var ButtonSet_1 = require("./ButtonSet");
var ButtonGroupFactory = react_1.forwardRef(function (_a, ref) {
    var onChange = _a.onChange, value = _a.value, props = __rest(_a, ["onChange", "value"]);
    var isControlled = utils_1.useControlWarn({
        controllingProps: ['onChange', 'value'],
        isControlledCheck: function () { return onChange !== undefined; },
        name: 'ButtonGroup',
    });
    function handleChange(e) {
        var newValue = xor_1.default(value, [e.target.value]);
        onChange && onChange(newValue);
    }
    return (react_1.default.createElement(ButtonSet_1.ButtonSet, __assign({}, props, { ref: ref }, (isControlled
        ? {
            onChange: handleChange,
            value: value,
        }
        : { name: props.name || uniqueId_1.default() }))));
});
exports.ButtonGroup = styled_components_1.default(ButtonGroupFactory).withConfig({ displayName: "ButtonGroup", componentId: "sc-chg3ap" })(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  ", " {\n    height: 36px;\n    border-radius: 4px;\n\n    + ", " {\n      margin-left: ", ";\n    }\n  }\n\n  ", " {\n    border-style: solid;\n    border-width: 1px;\n  }\n"], ["\n  ", " {\n    height: 36px;\n    border-radius: 4px;\n\n    + ", " {\n      margin-left: ", ";\n    }\n  }\n\n  ", " {\n    border-style: solid;\n    border-width: 1px;\n  }\n"])), ButtonItem_1.ButtonItem, ButtonItem_1.ButtonItem, function (props) { return props.theme.space.xxsmall; }, ButtonItem_1.ButtonItemLabel);
var templateObject_1;
