// @looker/components/dist/cjs/Button/size.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var styled_system_1 = require("styled-system");
exports.buttonSizeMap = {
    xxsmall: 18,
    xsmall: 22,
    small: 26,
    medium: 34,
    large: 42,
};
exports.buttonSize = styled_system_1.variant({
    prop: 'size',
    variants: {
        xxsmall: {
            fontSize: 'xsmall',
            height: '16px',
            px: 'xxsmall',
        },
        xsmall: {
            fontSize: 'xsmall',
            height: exports.buttonSizeMap.xsmall + "px",
            px: 'xsmall',
        },
        small: {
            fontSize: 'small',
            height: exports.buttonSizeMap.small + "px",
            px: 'small',
        },
        medium: {
            fontSize: 'medium',
            height: exports.buttonSizeMap.medium + "px",
            px: 'medium',
        },
        large: {
            fontSize: 'xlarge',
            height: exports.buttonSizeMap.large + "px",
            px: 'large',
        },
    },
});
