// @looker/components/dist/esm/Form/Inputs/ToggleSwitch/ToggleSwitch.js
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { rem, rgba } from 'polished';
import React, { forwardRef } from 'react';
import styled from 'styled-components';
import { palette, pseudoClasses, reset, space, } from '@looker/design-tokens';
export var CustomizableToggleSwitchAttributes = {
    knobOffColor: palette.white,
    knobOnColor: palette.white,
    offColor: palette.charcoal300,
    onColor: palette.purple300,
};
var Knob = styled(function (_a) {
    var className = _a.className;
    return React.createElement("div", { className: className });
}).withConfig({ displayName: "Knob", componentId: "sc-l93e8z" })(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  transform: ", ";\n  transition: ", ";\n  position: absolute;\n  bottom: ", ";\n  left: ", ";\n  width: ", ";\n  height: ", ";\n  border-radius: 50%;\n  background: ", ";\n"], ["\n  transform: ",
    ";\n  transition: ", ";\n  position: absolute;\n  bottom: ", ";\n  left: ", ";\n  width: ", ";\n  height: ", ";\n  border-radius: 50%;\n  background: ",
    ";\n"])), function (props) {
    return props.on ? "translateX(" + rem(props.size * 0.75) + ")" : '';
}, function (props) { return props.theme.transitions.durationModerate; }, function (props) { return rem(props.size * 0.1); }, function (props) { return rem(props.size * 0.1); }, function (props) { return rem(props.size * 0.8); }, function (props) { return rem(props.size * 0.8); }, function (props) {
    return props.on
        ? CustomizableToggleSwitchAttributes.knobOnColor
        : CustomizableToggleSwitchAttributes.knobOffColor;
});
var KnobContainer = function (_a) {
    var className = _a.className, props = __rest(_a, ["className"]);
    var hoverStyle = props.disabled
        ? undefined
        : { boxShadow: "0 0 .01rem 0.01rem " + rgba(palette.primary500, 0.5) };
    return (React.createElement(KnobContainerBase, { className: className, hoverStyle: hoverStyle, size: props.size, on: props.on },
        React.createElement(Knob, __assign({}, props))));
};
var KnobContainerBase = styled(function (_a) {
    var children = _a.children, className = _a.className;
    return (React.createElement("div", { className: className }, children));
}).withConfig({ displayName: "KnobContainerBase", componentId: "sc-66wsxs" })(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n  ", "\n  ", "\n\n  transition: ", ";\n  position: absolute;\n  top: 0;\n  bottom: 0;\n  left: 0;\n  right: 0;\n  border-radius: ", ";\n  background: ", ";\n"], ["\n  ", "\n  ", "\n\n  transition: ", ";\n  position: absolute;\n  top: 0;\n  bottom: 0;\n  left: 0;\n  right: 0;\n  border-radius: ", ";\n  background: ",
    ";\n"])), reset, pseudoClasses, function (props) { return props.theme.transitions.durationModerate; }, function (props) { return rem(props.size); }, function (props) {
    return props.on
        ? CustomizableToggleSwitchAttributes.onColor
        : CustomizableToggleSwitchAttributes.offColor;
});
var HiddenCheckbox = styled.input.attrs({ type: 'checkbox' }).withConfig({ displayName: "HiddenCheckbox", componentId: "sc-1oeh1g0" })(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n  opacity: 0;\n  width: 100%;\n  height: 100%;\n  position: absolute;\n  top: 0;\n  left: 0;\n  z-index: 1;\n"], ["\n  opacity: 0;\n  width: 100%;\n  height: 100%;\n  position: absolute;\n  top: 0;\n  left: 0;\n  z-index: 1;\n"])));
var DisabledKnob = styled.div.withConfig({ displayName: "DisabledKnob", componentId: "sc-1xddtn5" })(templateObject_4 || (templateObject_4 = __makeTemplateObject(["\n  ", "\n  position: absolute;\n  top: 0;\n  bottom: 0;\n  left: 0;\n  right: 0;\n  opacity: 0.4;\n  background: ", ";\n  border-radius: ", ";\n"], ["\n  ", "\n  position: absolute;\n  top: 0;\n  bottom: 0;\n  left: 0;\n  right: 0;\n  opacity: 0.4;\n  background: ", ";\n  border-radius: ", ";\n"])), reset, function (props) { return props.theme.colors.palette.charcoal300; }, function (props) { return rem(props.size); });
export var ToggleSwitchComponent = forwardRef(function (_a, ref) {
    var className = _a.className, disabled = _a.disabled, on = _a.on, _b = _a.size, size = _b === void 0 ? 20 : _b, props = __rest(_a, ["className", "disabled", "on", "size"]);
    return (React.createElement("div", { className: className },
        React.createElement(HiddenCheckbox, __assign({ checked: on, disabled: disabled, role: "switch", "aria-checked": on, ref: ref }, props)),
        React.createElement(KnobContainer, { size: size, on: on, disabled: disabled }),
        disabled && React.createElement(DisabledKnob, { size: size })));
});
ToggleSwitchComponent.displayName = 'ToggleSwitchComponent';
export var ToggleSwitch = styled(ToggleSwitchComponent).withConfig({ displayName: "ToggleSwitch", componentId: "sc-19bmztc" })(templateObject_5 || (templateObject_5 = __makeTemplateObject(["\n  ", "\n  input {\n    cursor: ", ";\n    &:focus + div {\n      box-shadow: 0 0 0 0.2rem ", ";\n    }\n  }\n  width: ", ";\n  height: ", ";\n  display: inline-block;\n  position: relative;\n  vertical-align: middle;\n  ", "\n"], ["\n  ", "\n  input {\n    cursor: ", ";\n    &:focus + div {\n      box-shadow: 0 0 0 0.2rem ", ";\n    }\n  }\n  width: ", ";\n  height: ", ";\n  display: inline-block;\n  position: relative;\n  vertical-align: middle;\n  ", "\n"])), reset, function (props) { return (!props.disabled ? 'pointer' : undefined); }, rgba(palette.primary500, 0.4), function (props) { return rem((props.size || 20) * 1.75); }, function (props) { return rem(props.size || 20); }, space);
var templateObject_1, templateObject_2, templateObject_3, templateObject_4, templateObject_5;
