// @looker/components/dist/esm/Form/Inputs/InputText/InputText.js
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import pick from 'lodash/pick';
import omit from 'lodash/omit';
import { border, typography, layout, pseudoClasses, space, reset, color, } from '@looker/design-tokens';
import React, { forwardRef } from 'react';
import styled from 'styled-components';
import { inputPropKeys } from '../InputProps';
export var CustomizableInputTextAttributes = {
    borderRadius: 'medium',
    fontSize: 'small',
    height: '28px',
    px: 'xsmall',
    py: 'none',
};
var InputComponent = forwardRef(function (_a, ref) {
    var _b = _a.type, type = _b === void 0 ? 'text' : _b, props = __rest(_a, ["type"]);
    return (React.createElement("input", __assign({}, pick(omit(props, 'height', 'width'), inputPropKeys), { type: type, className: props.className, ref: ref })));
});
InputComponent.displayName = 'InputComponent';
export var InputText = styled(InputComponent).attrs(function (props) { return ({
    background: props.validationType === 'error' && 'palette.red000',
    px: props.px || props.p || CustomizableInputTextAttributes.px,
    py: props.py || props.p || CustomizableInputTextAttributes.py,
}); }).withConfig({ displayName: "InputText", componentId: "sc-16mk1om" })(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  ", "\n  ", "\n  ", "\n  ", "\n  ", "\n  ", "\n  ", "\n"], ["\n  ", "\n  ", "\n  ", "\n  ", "\n  ", "\n  ", "\n  ", "\n"])), reset, border, color, layout, space, typography, pseudoClasses);
InputText.defaultProps = {
    border: 'solid 1px',
    borderColor: 'palette.charcoal300',
    borderRadius: CustomizableInputTextAttributes.borderRadius,
    fontSize: CustomizableInputTextAttributes.fontSize,
    height: CustomizableInputTextAttributes.height,
    type: 'text',
};
var templateObject_1;
