// @looker/components/dist/cjs/Form/Fields/FieldColor/ColorWheel/canvas_utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var d3_hsv_1 = require("d3-hsv");
var math_utils_1 = require("./math_utils");
exports.canvasMargin = 10;
exports.canvasRadius = function (canvas, margin) { return (canvas.width - 2 * margin) / 2; };
exports.eventCartesianPosition = function (canvas, event) { return ({
    x: math_utils_1.translate(-canvas.getBoundingClientRect().left, event.clientX),
    y: math_utils_1.translate(-canvas.getBoundingClientRect().top, event.clientY),
}); };
exports.clearCanvas = function (canvas) {
    var ctx = canvas.getContext('2d');
    if (ctx) {
        ctx.clearRect(0, 0, canvas.width, canvas.height);
    }
};
exports.drawColorWheelIntoCanvasImage = function (image, data) {
    data.forEach(function (row, x) {
        row.forEach(function (val, y) {
            var pixelWidth = 4;
            var index = (x + y * row.length) * pixelWidth;
            var rgbColor = d3_hsv_1.hsv(val.h, val.s, val.v).rgb();
            image[index] = rgbColor.r;
            image[index + 1] = rgbColor.g;
            image[index + 2] = rgbColor.b;
            image[index + 3] = 255;
        });
    });
};
