// @looker/components/dist/cjs/Form/Fields/FieldColor/ColorWheel/color_wheel_utils.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var range_1 = __importDefault(require("lodash/range"));
var math_utils_1 = require("./math_utils");
exports.white = function () { return ({
    brightness: 1,
    coord: { angle: 0, radius: 0 },
}); };
exports.hsv2polar = function (color) { return ({
    angle: math_utils_1.deg2rad(color.h),
    radius: color.s,
}); };
exports.polarbrightness2hsv = function (pb) { return ({
    h: math_utils_1.rad2deg(pb.coord.angle),
    s: pb.coord.radius,
    v: pb.brightness,
}); };
exports.scalePBRadius = function (by, pb) { return (__assign(__assign({}, pb), { coord: math_utils_1.scaleRadius(by, pb.coord) })); };
exports.cartesian2hsv = function (brightness, radius, coord) {
    return [coord]
        .map(function (c) { return math_utils_1.translateDiagonal(-radius, c); })
        .map(math_utils_1.cartesian2polar)
        .map(function (c) { return (c.radius <= radius ? { brightness: brightness, coord: c } : exports.white()); })
        .map(function (pb) { return exports.scalePBRadius(1 / radius, pb); })
        .map(exports.polarbrightness2hsv)[0];
};
exports.hsv2cartesian = function (radius, color) {
    return [color]
        .map(exports.hsv2polar)
        .map(function (coord) { return math_utils_1.scaleRadius(radius, coord); })
        .map(math_utils_1.polar2cartesian)
        .map(function (coord) { return math_utils_1.translateDiagonal(radius, coord); })[0];
};
exports.generateColorWheel = function (radius, brightness) {
    return range_1.default(math_utils_1.diameter(radius)).map(function (_, x) {
        return range_1.default(math_utils_1.diameter(radius)).map(function (_$, y) {
            return exports.cartesian2hsv(brightness, radius, { x: x, y: y });
        });
    });
};
