// @looker/components/dist/esm/Tooltip/Tooltip.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import React, { useRef, useState } from 'react';
import { Popper } from 'react-popper';
import { ModalContext } from '../Modal';
import { OverlaySurface } from '../Overlay/OverlaySurface';
import { Paragraph } from '../Text';
export var CustomizableTooltipAttributes = {};
export function useTooltip(_a) {
    var _b = _a.arrow, arrow = _b === void 0 ? true : _b, canClose = _a.canClose, content = _a.content, _c = _a.isOpen, initializeOpen = _c === void 0 ? false : _c, _d = _a.maxWidth, maxWidth = _d === void 0 ? '16rem' : _d, _e = _a.width, width = _e === void 0 ? 'auto' : _e, _f = _a.textAlign, textAlign = _f === void 0 ? 'center' : _f, disabled = _a.disabled, surfaceStyles = _a.surfaceStyles, props = __rest(_a, ["arrow", "canClose", "content", "isOpen", "maxWidth", "width", "textAlign", "disabled", "surfaceStyles"]);
    var _g = useState(initializeOpen), isOpen = _g[0], setIsOpen = _g[1];
    var surfaceRef = useRef(null);
    var newTriggerRef = useRef(null);
    var triggerRef = props.triggerRef || newTriggerRef;
    var handleOpen = function () { return setIsOpen(true); };
    var handleClose = function () {
        if (canClose && !canClose())
            return;
        setIsOpen(false);
    };
    var handleMouseOut = function (event) {
        if (!isOpen)
            return;
        var related = event.relatedTarget;
        if (triggerRef.current &&
            (triggerRef.current === related ||
                triggerRef.current.contains(related))) {
            return;
        }
        if (surfaceRef.current &&
            (surfaceRef.current === related ||
                surfaceRef.current.contains(related))) {
            return;
        }
        handleClose();
    };
    var eventHandlers = {
        onBlur: handleClose,
        onFocus: handleOpen,
        onMouseOut: handleMouseOut,
        onMouseOver: handleOpen,
    };
    var setSurfaceRef = function (ref) {
        surfaceRef.current = ref;
    };
    var referenceElement = triggerRef && triggerRef.current ? triggerRef.current : undefined;
    var contentFormatted = (React.createElement(Paragraph, { style: { hyphens: 'auto', overflowWrap: 'anywhere' }, fontSize: "xsmall", maxWidth: maxWidth, width: width, p: "xsmall", m: "none", textAlign: textAlign }, content));
    var popper = isOpen && content && !disabled ? (React.createElement(ModalContext.Provider, { value: { closeModal: handleClose } },
        React.createElement(Popper, { positionFixed: true, innerRef: setSurfaceRef, placement: props.placement, modifiers: {
                flip: {
                    behavior: 'flip',
                    enabled: true,
                    flipVariations: true,
                    flipVariationsByContent: true,
                },
                preventOverflow: {
                    padding: 0,
                },
            }, referenceElement: referenceElement }, function (_a) {
            var ref = _a.ref, style = _a.style, placement = _a.placement, arrowProps = _a.arrowProps;
            return (React.createElement(OverlaySurface, __assign({ arrow: arrow, arrowProps: arrowProps, eventHandlers: { onMouseOut: handleMouseOut }, placement: placement, ref: ref, style: style, zIndex: CustomizableTooltipAttributes.zIndex, backgroundColor: "palette.charcoal600", borderRadius: "medium", boxShadow: 3, color: "palette.charcoal000" }, surfaceStyles), contentFormatted));
        }))) : null;
    return {
        eventHandlers: eventHandlers,
        ref: triggerRef,
        tooltip: popper,
    };
}
export var Tooltip = function (_a) {
    var children = _a.children, props = __rest(_a, ["children"]);
    var _b = useTooltip(props), eventHandlers = _b.eventHandlers, tooltip = _b.tooltip, ref = _b.ref;
    return (React.createElement(React.Fragment, null,
        tooltip,
        children(eventHandlers, ref)));
};
