// @looker/components/dist/cjs/Form/Fieldset/Legend.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var styled_components_1 = __importDefault(require("styled-components"));
var design_tokens_1 = require("@looker/design-tokens");
exports.CustomizableLegendAttributes = {
    bottomPadding: 'xsmall',
    color: 'palette.charcoal800',
    fontSize: 'xxxlarge',
    fontWeight: 'light',
};
exports.Legend = styled_components_1.default.legend.attrs(function (props) { return ({
    color: props.color || exports.CustomizableLegendAttributes.color,
    fontSize: props.fontSize || exports.CustomizableLegendAttributes.fontSize,
    fontWeight: props.fontWeight || exports.CustomizableLegendAttributes.fontWeight,
    pb: props.pb || props.py || props.p
        ? undefined
        : exports.CustomizableLegendAttributes.bottomPadding,
}); }).withConfig({ displayName: "Legend", componentId: "sc-oh7l4v" })(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  ", "\n  ", "\n  ", "\n  ", "\n  ", ";\n  ", "\n"], ["\n  ", "\n  ", "\n  ", "\n  ", "\n  ", ";\n  ", "\n"])), design_tokens_1.reset, design_tokens_1.color, design_tokens_1.layout, design_tokens_1.space, design_tokens_1.textTransform, design_tokens_1.typography);
var templateObject_1;
