// @looker/components/dist/cjs/Divider/Divider.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var design_tokens_1 = require("@looker/design-tokens");
var styled_components_1 = __importDefault(require("styled-components"));
var styled_system_1 = require("styled-system");
var appearanceVariant = styled_system_1.variant({
    prop: 'appearance',
    variants: {
        dark: {
            bg: 'palette.charcoal400',
        },
        default: {
            bg: 'palette.charcoal300',
        },
        light: {
            bg: 'palette.charcoal200',
        },
        onDark: {
            bg: 'palette.charcoal500',
        },
    },
});
exports.Divider = styled_components_1.default.hr.attrs(function (props) { return ({
    bg: props.customColor,
}); }).withConfig({ displayName: "Divider", componentId: "sc-1hvv94c" })(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  ", "\n  ", "\n  ", "\n  ", "\n\n  height: ", ";\n  ", "\n"], ["\n  ", "\n  ", "\n  ", "\n  ", "\n\n  height: ", ";\n  ", "\n"])), design_tokens_1.reset, design_tokens_1.position, design_tokens_1.border, design_tokens_1.space, function (props) { return props.size; }, function (props) { return (props.customColor ? design_tokens_1.color : appearanceVariant); });
exports.Divider.defaultProps = { appearance: 'default', size: '1px' };
var templateObject_1;
