// @looker/design-tokens/dist/esm/tokens/color/semantic_colors.js
import { palette } from './palette';
var blue500 = palette.blue500, charcoal000 = palette.charcoal000, charcoal100 = palette.charcoal100, charcoal300 = palette.charcoal300, charcoal400 = palette.charcoal400, charcoal500 = palette.charcoal500, charcoal600 = palette.charcoal600, primary500 = palette.primary500, primary600 = palette.primary600, primary700 = palette.primary700, purple000 = palette.purple000, purple100 = palette.purple100, purple400 = palette.purple400, red000 = palette.red000, red100 = palette.red100, red400 = palette.red400, red500 = palette.red500, red600 = palette.red600, red700 = palette.red700, white = palette.white;
export var semanticColors = {
    danger: {
        altText: red400,
        borderColor: charcoal300,
        dark: red600,
        darker: red700,
        light: red100,
        lighter: red000,
        linkColor: blue500,
        main: red500,
        text: white,
    },
    neutral: {
        altText: charcoal600,
        borderColor: charcoal300,
        dark: charcoal500,
        darker: charcoal600,
        light: charcoal100,
        lighter: charcoal000,
        linkColor: blue500,
        main: charcoal400,
        text: white,
    },
    primary: {
        altText: purple400,
        borderColor: charcoal300,
        dark: primary600,
        darker: primary700,
        light: purple100,
        lighter: purple000,
        linkColor: blue500,
        main: primary500,
        text: white,
    },
    secondary: {
        altText: purple400,
        borderColor: charcoal300,
        dark: primary600,
        darker: primary700,
        light: purple100,
        lighter: purple000,
        linkColor: blue500,
        main: primary500,
        text: white,
    },
};
