// @looker/design-tokens/dist/esm/system/reset.js
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
import { css } from 'styled-components';
export var defaultReset = function () {
    var simpleReset = {
        border: 0,
        boxSizing: 'border-box',
        font: 'inherit',
        fontSize: '100%',
        margin: 0,
        padding: 0,
        verticalAlign: 'baseline',
    };
    return simpleReset;
};
export var reset = css(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  /**\n     * Global reset applied to prevent styling on top level tags outside of Looker \n     * UI Components from interfering with styles.\n     *\n     * This **must** be first.\n     */\n  ", "\n"], ["\n  /**\n     * Global reset applied to prevent styling on top level tags outside of Looker \n     * UI Components from interfering with styles.\n     *\n     * This **must** be first.\n     */\n  ", "\n"])), function (props) { return props.theme.reset && props.theme.reset(); });
var templateObject_1;
