// @looker/components/dist/esm/List/List.js
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
import { reset, space, typography, position, layout, } from '@looker/design-tokens';
import styled from 'styled-components';
import { variant } from 'styled-system';
var typeVariant = variant({
    prop: 'type',
    variants: {
        bullet: {
            listStyleType: 'disc',
            pl: 'medium',
        },
        letter: {
            listStyleType: 'upper-alpha',
            pl: 'medium',
        },
        none: {
            listStyleType: 'none',
        },
        number: {
            listStyleType: 'decimal',
            pl: 'medium',
        },
    },
});
export var List = styled.ul.attrs(function (props) { return ({
    as: ['letter', 'number'].includes(String(props.type)) ? 'ol' : undefined,
    type: props.nomarker ? 'none' : props.type,
}); }).withConfig({ displayName: "List", componentId: "sc-1dd290l" })(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  ", "\n  ", "\n  ", "\n  ", "\n\n  ", "\n  ", "\n"], ["\n  ", "\n  ", "\n  ", "\n  ", "\n\n  ", "\n  ", "\n"])), reset, typography, typeVariant, space, position, layout);
List.defaultProps = { type: 'none' };
var templateObject_1;
