// @looker/design-tokens/dist/cjs/tokens/color/semantic_colors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var palette_1 = require("./palette");
var blue500 = palette_1.palette.blue500, charcoal000 = palette_1.palette.charcoal000, charcoal100 = palette_1.palette.charcoal100, charcoal300 = palette_1.palette.charcoal300, charcoal400 = palette_1.palette.charcoal400, charcoal500 = palette_1.palette.charcoal500, charcoal600 = palette_1.palette.charcoal600, primary500 = palette_1.palette.primary500, primary600 = palette_1.palette.primary600, primary700 = palette_1.palette.primary700, purple000 = palette_1.palette.purple000, purple100 = palette_1.palette.purple100, purple400 = palette_1.palette.purple400, red000 = palette_1.palette.red000, red100 = palette_1.palette.red100, red400 = palette_1.palette.red400, red500 = palette_1.palette.red500, red600 = palette_1.palette.red600, red700 = palette_1.palette.red700, white = palette_1.palette.white;
exports.semanticColors = {
    danger: {
        altText: red400,
        borderColor: charcoal300,
        dark: red600,
        darker: red700,
        light: red100,
        lighter: red000,
        linkColor: blue500,
        main: red500,
        text: white,
    },
    neutral: {
        altText: charcoal600,
        borderColor: charcoal300,
        dark: charcoal500,
        darker: charcoal600,
        light: charcoal100,
        lighter: charcoal000,
        linkColor: blue500,
        main: charcoal400,
        text: white,
    },
    primary: {
        altText: purple400,
        borderColor: charcoal300,
        dark: primary600,
        darker: primary700,
        light: purple100,
        lighter: purple000,
        linkColor: blue500,
        main: primary500,
        text: white,
    },
    secondary: {
        altText: purple400,
        borderColor: charcoal300,
        dark: primary600,
        darker: primary700,
        light: purple100,
        lighter: purple000,
        linkColor: blue500,
        main: primary500,
        text: white,
    },
};
