// @looker/components/dist/esm/Form/Fields/FieldColor/ColorWheel/canvas_utils.js
import { hsv } from 'd3-hsv';
import { translate } from './math_utils';
export var canvasMargin = 10;
export var canvasRadius = function (canvas, margin) { return (canvas.width - 2 * margin) / 2; };
export var eventCartesianPosition = function (canvas, event) { return ({
    x: translate(-canvas.getBoundingClientRect().left, event.clientX),
    y: translate(-canvas.getBoundingClientRect().top, event.clientY),
}); };
export var clearCanvas = function (canvas) {
    var ctx = canvas.getContext('2d');
    if (ctx) {
        ctx.clearRect(0, 0, canvas.width, canvas.height);
    }
};
export var drawColorWheelIntoCanvasImage = function (image, data) {
    data.forEach(function (row, x) {
        row.forEach(function (val, y) {
            var pixelWidth = 4;
            var index = (x + y * row.length) * pixelWidth;
            var rgbColor = hsv(val.h, val.s, val.v).rgb();
            image[index] = rgbColor.r;
            image[index + 1] = rgbColor.g;
            image[index + 2] = rgbColor.b;
            image[index + 3] = 255;
        });
    });
};
