// @looker/components/dist/cjs/Modal/ModalSurface.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var design_tokens_1 = require("@looker/design-tokens");
var react_1 = __importStar(require("react"));
var react_hotkeys_1 = require("react-hotkeys");
var styled_components_1 = __importStar(require("styled-components"));
var styled_system_1 = require("styled-system");
var utils_1 = require("../utils");
var ModalContext_1 = require("./ModalContext");
exports.ModalSurface = function (_a) {
    var anchor = _a.anchor, style = _a.style, className = _a.className, props = __rest(_a, ["anchor", "style", "className"]);
    var closeModal = react_1.useContext(ModalContext_1.ModalContext).closeModal;
    var _b = react_1.useState(false), focusTrapEnabled = _b[0], setFocusTrapEnabled = _b[1];
    var focusRef = utils_1.useFocusTrap(focusTrapEnabled);
    react_1.useEffect(function () {
        window.setTimeout(function () {
            setFocusTrapEnabled(true);
        }, design_tokens_1.theme.transitions.durationModerate);
    }, []);
    return (react_1.default.createElement(react_hotkeys_1.HotKeys, { keyMap: {
            CLOSE_MODAL: {
                action: 'keyup',
                name: 'Close Modal',
                sequence: 'esc',
            },
        }, handlers: {
            CLOSE_MODAL: function () {
                closeModal && closeModal();
            },
        }, style: {
            alignItems: 'center',
            display: 'flex',
            height: '100%',
            justifyContent: anchor === 'right' ? 'flex-end' : 'center',
            width: '100%',
        } },
        react_1.default.createElement(Style, __assign({ className: "surface-overflow " + className, tabIndex: -1, surfaceStyle: style, ref: focusRef }, props))));
};
var surfaceTransition = function () { return styled_components_1.css(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  ", "\n"], ["\n  ",
    "\n"])), function (props) {
    return props.theme.transitions.durationModerate + " " + props.theme.easings.ease;
}); };
var Style = styled_components_1.default.div.withConfig({ displayName: "Style", componentId: "sc-dsrm29" })(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n  ", "\n  ", "\n  ", "\n  ", "\n\n  ", "\n\n  display: flex;\n  flex-direction: column;\n  max-width: 100%;\n  position: relative;\n  transition: transform ", ", opacity ", ";\n\n  ", ";\n\n  &:focus {\n    outline: none;\n  }\n"], ["\n  ", "\n  ", "\n  ", "\n  ", "\n\n  ", "\n\n  display: flex;\n  flex-direction: column;\n  max-width: 100%;\n  position: relative;\n  transition: transform ", ", opacity ", ";\n\n  ", ";\n\n  &:focus {\n    outline: none;\n  }\n"])), design_tokens_1.reset, styled_system_1.boxShadow, styled_system_1.border, styled_system_1.layout, styled_system_1.color, surfaceTransition, surfaceTransition, function (props) { return props.surfaceStyle; });
Style.defaultProps = {
    backgroundColor: 'palette.white',
    borderRadius: 'medium',
    boxShadow: 3,
    color: 'palette.charcoal900',
};
var templateObject_1, templateObject_2;
