// @looker/components/dist/es2015/Modal/ModalSurface.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { reset, theme } from '@looker/design-tokens';
import React, { useContext, useEffect, useState } from 'react';
import { HotKeys } from 'react-hotkeys';
import styled, { css } from 'styled-components';
import { boxShadow, border, color, layout, } from 'styled-system';
import { useFocusTrap } from '../utils';
import { ModalContext } from './ModalContext';
export const ModalSurface = (_a) => {
    var { anchor, style, className } = _a, props = __rest(_a, ["anchor", "style", "className"]);
    const { closeModal } = useContext(ModalContext);
    const [focusTrapEnabled, setFocusTrapEnabled] = useState(false);
    const focusRef = useFocusTrap(focusTrapEnabled);
    useEffect(() => {
        window.setTimeout(() => {
            setFocusTrapEnabled(true);
        }, theme.transitions.durationModerate);
    }, []);
    return (React.createElement(HotKeys, { keyMap: {
            CLOSE_MODAL: {
                action: 'keyup',
                name: 'Close Modal',
                sequence: 'esc',
            },
        }, handlers: {
            CLOSE_MODAL: () => {
                closeModal && closeModal();
            },
        }, style: {
            alignItems: 'center',
            display: 'flex',
            height: '100%',
            justifyContent: anchor === 'right' ? 'flex-end' : 'center',
            width: '100%',
        } },
        React.createElement(Style, Object.assign({ className: `surface-overflow ${className}`, tabIndex: -1, surfaceStyle: style, ref: focusRef }, props))));
};
const surfaceTransition = () => css `
  ${props => `${props.theme.transitions.durationModerate} ${props.theme.easings.ease}`}
`;
const Style = styled.div.withConfig({ displayName: "Style", componentId: "sc-dsrm29" }) `
  ${reset}
  ${boxShadow}
  ${border}
  ${layout}

  ${color}

  display: flex;
  flex-direction: column;
  max-width: 100%;
  position: relative;
  transition: transform ${surfaceTransition}, opacity ${surfaceTransition};

  ${props => props.surfaceStyle};

  &:focus {
    outline: none;
  }
`;
Style.defaultProps = {
    backgroundColor: 'palette.white',
    borderRadius: 'medium',
    boxShadow: 3,
    color: 'palette.charcoal900',
};
