// @looker/components/dist/es2015/Modal/ModalPortal.js
import React, { useEffect, useRef } from 'react';
import { createPortal } from 'react-dom';
import styled from 'styled-components';
import { useScrollLock } from '../utils/useScrollLock';
import { CustomizableModalAttributes } from './Modal';
import { getModalRoot } from './modalRoot';
export const ModalPortal = ({ portalRef, children }) => {
    const el = useRef(document.createElement('div'));
    const ref = useRef(null);
    const refToUse = portalRef || ref;
    useScrollLock(refToUse);
    useEffect(() => {
        const modalRoot = getModalRoot();
        if (!modalRoot)
            return;
        const elCurrent = el.current;
        modalRoot.appendChild(elCurrent);
        return () => {
            modalRoot.removeChild(elCurrent);
        };
    }, [el]);
    const content = (React.createElement(InvisiBox, { ref: refToUse, zIndex: CustomizableModalAttributes.zIndex }, children));
    return createPortal(content, el.current);
};
const InvisiBox = styled.div.withConfig({ displayName: "InvisiBox", componentId: "sc-1ali4oo" }) `
  position: fixed;
  bottom: 0;
  left: 0;
  right: 0;
  top: 0;

  align-items: center;
  display: flex;
  justify-content: center;
  pointer-events: none;
  z-index: ${props => props.zIndex};

  * {
    pointer-events: auto;
  }
`;
