// @looker/components/dist/cjs/List/List.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var design_tokens_1 = require("@looker/design-tokens");
var styled_components_1 = __importDefault(require("styled-components"));
var styled_system_1 = require("styled-system");
var typeVariant = styled_system_1.variant({
    prop: 'type',
    variants: {
        bullet: {
            listStyleType: 'disc',
            pl: 'medium',
        },
        letter: {
            listStyleType: 'upper-alpha',
            pl: 'medium',
        },
        none: {
            listStyleType: 'none',
        },
        number: {
            listStyleType: 'decimal',
            pl: 'medium',
        },
    },
});
exports.List = styled_components_1.default.ul.attrs(function (props) { return ({
    as: ['letter', 'number'].includes(String(props.type)) ? 'ol' : undefined,
    type: props.nomarker ? 'none' : props.type,
}); }).withConfig({ displayName: "List", componentId: "sc-1dd290l" })(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  ", "\n  ", "\n  ", "\n  ", "\n\n  ", "\n  ", "\n"], ["\n  ", "\n  ", "\n  ", "\n  ", "\n\n  ", "\n  ", "\n"])), design_tokens_1.reset, design_tokens_1.typography, typeVariant, design_tokens_1.space, design_tokens_1.position, design_tokens_1.layout);
exports.List.defaultProps = { type: 'none' };
var templateObject_1;
