// @looker/components/dist/cjs/Form/Inputs/InputText/InputText.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var pick_1 = __importDefault(require("lodash/pick"));
var omit_1 = __importDefault(require("lodash/omit"));
var design_tokens_1 = require("@looker/design-tokens");
var react_1 = __importStar(require("react"));
var styled_components_1 = __importDefault(require("styled-components"));
var InputProps_1 = require("../InputProps");
exports.CustomizableInputTextAttributes = {
    borderRadius: 'medium',
    fontSize: 'small',
    height: '28px',
    px: 'xsmall',
    py: 'none',
};
var InputComponent = react_1.forwardRef(function (_a, ref) {
    var _b = _a.type, type = _b === void 0 ? 'text' : _b, props = __rest(_a, ["type"]);
    return (react_1.default.createElement("input", __assign({}, pick_1.default(omit_1.default(props, 'height', 'width'), InputProps_1.inputPropKeys), { type: type, className: props.className, ref: ref })));
});
InputComponent.displayName = 'InputComponent';
exports.InputText = styled_components_1.default(InputComponent).attrs(function (props) { return ({
    background: props.validationType === 'error' && 'palette.red000',
    px: props.px || props.p || exports.CustomizableInputTextAttributes.px,
    py: props.py || props.p || exports.CustomizableInputTextAttributes.py,
}); }).withConfig({ displayName: "InputText", componentId: "sc-16mk1om" })(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  ", "\n  ", "\n  ", "\n  ", "\n  ", "\n  ", "\n  ", "\n"], ["\n  ", "\n  ", "\n  ", "\n  ", "\n  ", "\n  ", "\n  ", "\n"])), design_tokens_1.reset, design_tokens_1.border, design_tokens_1.color, design_tokens_1.layout, design_tokens_1.space, design_tokens_1.typography, design_tokens_1.pseudoClasses);
exports.InputText.defaultProps = {
    border: 'solid 1px',
    borderColor: 'palette.charcoal300',
    borderRadius: exports.CustomizableInputTextAttributes.borderRadius,
    fontSize: exports.CustomizableInputTextAttributes.fontSize,
    height: exports.CustomizableInputTextAttributes.height,
    type: 'text',
};
var templateObject_1;
